cpp/bipartite_matching_module/algorithm/bipartite_matching.cpp

```diff
@@ -0,0 +1,122 @@
+#include <queue>
+#include <set>
+
+#include "bipartite_matching.hpp"
+
+namespace bipartite_matching_util {
+bool BipartiteMatchingDFS(const std::uint64_t node, const std::vector<std::vector<std::uint64_t>> &adj_list,
+                          std::vector<bool> &visited, std::vector<std::optional<std::uint64_t>> &matched) {
+  if (visited[node]) return false;
+
+  visited[node] = true;
+  for (const auto next : adj_list[node]) {
+    if (const auto matched_next = matched[next];
+        !matched_next || BipartiteMatchingDFS(*matched_next, adj_list, visited, matched)) {
+      matched[next] = node;
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsGraphBipartite(const mg_graph::GraphView<> &graph) {
+  auto node_size = graph.Nodes().size();
+
+  // -1 to indicate that color is not set to that node
+  std::vector<std::int8_t> colors(node_size, -1);
+
+  for (std::uint64_t i = 0; i < node_size; i++) {
+    if (colors[i] == -1) {
+      if (!IsSubgraphBipartite(graph, colors, i)) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+bool IsSubgraphBipartite(const mg_graph::GraphView<> &graph, std::vector<std::int8_t> &colors,
+                         const std::uint64_t node_index) {
+  // Data structure used in BFS
+  std::queue<std::uint64_t> unvisited;
+
+  colors[node_index] = 1;
+  unvisited.push(node_index);
+  while (!unvisited.empty()) {
+    auto current_index = unvisited.front();
+    unvisited.pop();
+    for (const auto &neighbour : graph.Neighbours(current_index)) {
+      // Self loops are not allowed
+      if (neighbour.node_id == current_index) return false;
+
+      // If neighbor has the same color as the current node, graph is not bipartite
+      if (colors[neighbour.node_id] == colors[current_index]) return false;
+
+      if (colors[neighbour.node_id] == -1) {
+        // Set to the opposite color = 0/1
+        colors[neighbour.node_id] = 1 - colors[current_index];
+        unvisited.push(neighbour.node_id);
+      }
+    }
+  }
+
+  return true;
+}
+
+std::uint64_t MaximumMatching(const std::vector<std::pair<std::uint64_t, std::uint64_t>> &bipartite_edges) {
+  std::set<std::uint64_t> group_a;
+  std::set<std::uint64_t> group_b;
```

Comment:
Is the order of the nodes important?

Reply:
It's not, therefore I am adding unordered_set and unordered_map, where suitable